File: hippius_s3/writer/landed.py
"""Announce a freshly-written part to the node's drain agent.

The drain agent's reconciler used to be the sole way a new part was discovered: it walked the
whole SSD cache every 15s and asked the database which parts it had never seen. That was cheap
while the SSD held only the undrained backlog, because a drained part was unlinked. Retention
broke it — the disk now holds the node's entire replicated shard, measured on prod 2026-08-07 at
2.28M parts / ~930 GB, so the walk became millions of stat calls and thousands of batched
queries every 15s per node to find a handful of new parts.

Announcing turns discovery into O(new parts). The agent consumes this queue
(`crates/hippius-drain-agent/src/landed.rs`) and writes the `cephor_replication_status` row
itself — the api never touches that table, which belongs to the drain and to the drain's
migrations.

**Best-effort by construction.** A dropped announcement costs latency and nothing else: the
reconciler still finds the part on disk, exactly as it does today. That is what makes it safe to
swallow every failure here rather than fail a PUT that has already durably written its data.

Wiring follows the repo's module-singleton pattern (`get_access_tracker`):
`initialize_landed_publisher` in the api lifespan; processes that never initialize it — workers,
scripts, tests — get `None` and publishing is a no-op.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import TYPE_CHECKING
from typing import Any
from typing import Optional


if TYPE_CHECKING:
    from hippius_s3.monitoring import LandedAnnounceOutcome


logger = logging.getLogger(__name__)


# Must match `landed_queue_key` in crates/hippius-drain-agent/src/landed.rs.
_LANDED_QUEUE_PREFIX = "cephor:landed:"

# Bounds the one await this puts on the client PUT path.
#
# Sized to cover a TCP CONNECT plus a round trip, not just a round trip. redis-py drops the
# pooled connection when a command is cancelled, so the publish after a timeout has to reconnect
# inside this same budget — at 1s that was self-reinforcing, and because redis latency is a
# shared property every api pod crossed the threshold at the same moment, losing 100% of
# announcements fleet-wide rather than a sampled fraction. 5s keeps the bound (an unbounded await
# on a client with no socket_timeout was the actual defect) while putting it far outside the
# range a merely-loaded queue reaches.
#
# A timeout is NOT free, which is why it is counted rather than only logged: see
# `landed_announce_failures_total` and the note on the counter in monitoring.py.
_PUBLISH_TIMEOUT_SECONDS = 5.0

# Cap on a node's queue. The agent normally drains this within a poll, so depth is near zero;
# the bound only matters when the agent is down. Past it the OLDEST announcements are dropped,
# which is safe precisely because the reconciler backstop still finds those parts on disk — and
# is preferable to letting an unbounded list grow on a 1 GB Redis while the agent is away.
_DEFAULT_MAX_QUEUE_DEPTH = 200_000


def landed_queue_key(node_name: str) -> str:
    return f"{_LANDED_QUEUE_PREFIX}{node_name}"


def _record_announce_failure(outcome: "LandedAnnounceOutcome") -> None:
    """Best-effort counter bump; metrics must never be the reason a PUT fails."""
    try:
        from hippius_s3.monitoring import get_metrics_collector

        collector = get_metrics_collector()
        if collector is not None:
            collector.record_landed_announce_failure(outcome)
    except Exception:  # noqa: BLE001 - a metrics fault must not surface on the write path
        logger.debug("recording a landed-announce failure failed", exc_info=True)


class LandedPartPublisher:
    """Publishes "this node just finished writing a part" to its drain agent."""

    def __init__(self, queues_client: Any, node_name: str, max_depth: int = _DEFAULT_MAX_QUEUE_DEPTH) -> None:
        self._redis = queues_client
        self._key = landed_queue_key(node_name)
        self._max_depth = int(max_depth)

    async def publish(self, object_id: str, object_version: int, part_number: int) -> None:
        """Announce one completed part. Never raises.

        Field names are the wire contract with the agent's `LandedMessage`; the agent discards
        anything it cannot parse, so renaming one side alone silently disables the fast path
        rather than erroring. `drain_landed_dropped_total` is the alert for that.
        """
        payload = json.dumps(
            {
                "object_id": str(object_id),
                "version": int(object_version),
                "part_number": int(part_number),
            }
        )
        try:
            # LPUSH + LTRIM in one round-trip: the trim bounds the queue for the case where the
            # agent is down, and paying two round-trips per part on the write path to bound a
            # queue that is normally empty would be a poor trade.
            pipe = self._redis.pipeline()
            pipe.lpush(self._key, payload)
            pipe.ltrim(self._key, 0, self._max_depth - 1)
            # Bounded, because this await is on the CLIENT PUT PATH and the api's queues client is
            # built with no socket_timeout — so `except` below catches a redis-queues that ERRORS
            # and does nothing at all for one that merely goes SLOW. That is not hypothetical on
            # this queue: a 1.29M-entry list once made redis-queues slow enough to surface as
            # prod GET IncompleteRead. `promote_floor` bounds its own queues read for exactly this
            # reason; the write path is the one that needed it more. On timeout the part is
            # already durable and the agent's reconcile walk still finds it.
            await asyncio.wait_for(pipe.execute(), timeout=_PUBLISH_TIMEOUT_SECONDS)
        except TimeoutError:
            # WARNING and counted, not DEBUG. The module docstring's "a dropped announcement costs
            # latency and nothing else" is true only for DISCOVERY of a new part — the reconciler
            # does find that on disk. It is false for a RE-uploaded one: the reconciler tallies an
            # already-`replicated` part as an orphan and deliberately does not content-check it, so
            # the announcement is the only thing that triggers the divergence check. Losing one for
            # a rewritten part leaves the pool serving the previous attempt's ciphertext under the
            # new ETag, decrypting cleanly, with nothing else to notice.
            _record_announce_failure("timeout")
            logger.warning(
                "announcing landed part %s v%s part %s timed out after %ss; a rewrite of an already-"
                "replicated part would not be re-driven",
                object_id,
                object_version,
                part_number,
                _PUBLISH_TIMEOUT_SECONDS,
            )
        except Exception as exc:  # noqa: BLE001 - the part is already durable; the backstop covers this
            _record_announce_failure("error")
            logger.warning(
                "announcing landed part %s v%s part %s failed: %s", object_id, object_version, part_number, exc
            )


_publisher: Optional[LandedPartPublisher] = None


def initialize_landed_publisher(queues_client: Any, node_name: str) -> Optional[LandedPartPublisher]:
    """Installs the process-wide publisher, or `None` without a node identity.

    No `NODE_NAME` means no way to say WHICH node's agent should drain the part, and an
    announcement on the wrong queue would be recorded against a node that does not hold the
    data — where the node-scoped `claim_part` would then never drain it. Publishing nothing is
    strictly better: the reconciler on the node that DOES hold it still finds it.
    """
    global _publisher
    _publisher = LandedPartPublisher(queues_client, node_name) if queues_client is not None and node_name else None
    return _publisher


def get_landed_publisher() -> Optional[LandedPartPublisher]:
    return _publisher


__all__ = [
    "LandedPartPublisher",
    "get_landed_publisher",
    "initialize_landed_publisher",
    "landed_queue_key",
]

File: crates/hippius-drain-agent/src/landed.rs
//! The api's landed-part announcements: a per-node Redis queue the api pushes to when it
//! finishes writing a part, so the drain learns about it without walking the disk.
//!
//! # Why this exists
//!
//! Until now the reconciler was the SOLE discovery mechanism: every 15 s it walked the whole
//! SSD cache, materialised every part it found, and asked the database which of them it had
//! never seen. That was cheap while the SSD held only the undrained backlog — a drained part
//! was unlinked, so the tree was small. Retention broke that assumption: the disk now holds the
//! node's entire replicated shard, measured on prod 2026-08-07 at **2.28 M parts / ~930 GB per
//! node**, so the walk became ~2.28 M `stat` calls and ~4,560 batched Postgres round-trips
//! *every 15 seconds, per node*, to discover a handful of new parts.
//!
//! The announcement makes discovery O(new parts). The reconciler stays, demoted to a backstop
//! on a long poll: it is what recovers a part whose announcement was lost.
//!
//! # Why a queue and not a direct write from the api
//!
//! `cephor_replication_status` is the drain's table, owned by the drain's migrations. Letting
//! the api `INSERT` into it directly would give one table two writers across two services with
//! two independent migration systems. The repo already settled this shape for the mirror-image
//! handoff — the drain is the sole producer of upload requests and publishes them over Redis
//! rather than writing the consumer's state — so this follows it: the api announces, the agent
//! records.
//!
//! # Delivery guarantees, and why weak ones are enough
//!
//! At-most-once, deliberately. A lost message (Redis restart, a trimmed queue, an agent down
//! longer than the queue bound) costs latency and nothing else: the reconciler's next sweep
//! finds the part and records it exactly as it does today. So this is a fast path over a
//! correct-but-slow one, never a replacement for it — which is what lets the api treat its
//! publish as best-effort and lets this consumer drop a malformed message without ceremony.

use core::str::FromStr;
use hippius_drain_core::{ObjectId, PartKey, PartNumber, Version};
use redis::AsyncCommands;
use redis::aio::ConnectionManager;
use serde::Deserialize;
use thiserror::Error;

/// Redis key holding one node's landed-part announcements. Per node, because a part is only
/// ever on the SSD of the node that ingested it — a peer could not act on the message.
#[must_use]
pub fn landed_queue_key(node: &str) -> String {
    format!("cephor:landed:{node}")
}

/// A failure draining the announcement queue.
#[derive(Debug, Error)]
#[non_exhaustive]
pub enum LandedError {
    /// The Redis pop failed. Transient; the next poll retries and the reconciler backstops.
    #[error("reading the landed-part queue failed")]
    Redis(#[from] redis::RedisError),
}

/// One announcement as the api publishes it.
///
/// Field names are the wire contract with `hippius_s3/writer/landed.py`; changing either side
/// alone silently drops every message, since a parse failure is discarded by design.
#[derive(Debug, Deserialize)]
struct LandedMessage {
    object_id: String,
    version: u32,
    part_number: u32,
}

impl LandedMessage {
    /// The typed key, or `None` when the payload does not describe a real part.
    ///
    /// A malformed announcement is dropped rather than retried: it can only come from a
    /// version skew or a corrupted value, neither of which a retry fixes, and the reconciler
    /// still discovers the part from the disk.
    fn into_part(self) -> Option<PartKey> {
        let object = ObjectId::from_str(&self.object_id).ok()?;
        Some(PartKey::new(object, Version::new(self.version), PartNumber::new(self.part_number)))
    }
}

/// Pops landed-part announcements for one node.
///
/// Cheap to clone — the `ConnectionManager` is a shared multiplexed handle.
#[derive(Clone)]
pub struct LandedQueue {
    redis: ConnectionManager,
    key: String,
}

// Manual Debug: redis's `ConnectionManager` is not `Debug`, so derive cannot apply.
impl core::fmt::Debug for LandedQueue {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("LandedQueue").field("key", &self.key).finish_non_exhaustive()
    }
}

impl LandedQueue {
    /// Binds a queue to `node`'s announcement key.
    #[must_use]
    pub fn new(redis: ConnectionManager, node: &str) -> Self {
        Self {
            redis,
            key: landed_queue_key(node),
        }
    }

    /// Pops up to `limit` announcements, oldest first.
    ///
    /// `RPOP key count` takes the tail, which is the oldest end of an `LPUSH`ed list — so the
    /// queue is FIFO and a burst is drained in arrival order rather than newest-first.
    ///
    /// Unparseable entries are counted, not returned: see [`LandedMessage::into_part`].
    ///
    /// # Errors
    ///
    /// [`LandedError::Redis`] if the pop fails. Nothing is lost — the entries stay queued.
    pub async fn pop(&self, limit: usize) -> Result<(Vec<PartKey>, u64), LandedError> {
        let mut conn = self.redis.clone();
        let raw: Vec<String> = conn.rpop(&self.key, core::num::NonZeroUsize::new(limit)).await?;
        let mut parts = Vec::with_capacity(raw.len());
        let mut dropped = 0u64;
        for entry in raw {
            match serde_json::from_str::<LandedMessage>(&entry).ok().and_then(LandedMessage::into_part) {
                Some(part) => parts.push(part),
                None => dropped = dropped.saturating_add(1),
            }
        }
        Ok((parts, dropped))
    }
}

#[cfg(test)]
#[expect(clippy::unwrap_used, reason = "tests")]
mod tests {
    use super::{LandedMessage, landed_queue_key};

    const UUID: &str = "466916c0-d61b-4518-b81b-9576b574270a";

    #[test]
    fn the_queue_is_scoped_to_one_node() {
        // A part lives only on the SSD of the node that ingested it, so a peer receiving the
        // announcement could record a row pointing at a disk that does not hold the data —
        // and `claim_part` is node-scoped, so that part would never drain at all.
        assert_eq!(landed_queue_key("k8s-v3-node1"), "cephor:landed:k8s-v3-node1");
        assert_ne!(landed_queue_key("k8s-v3-node1"), landed_queue_key("k8s-v3-node2"));
    }

    #[test]
    fn a_well_formed_announcement_becomes_a_part_key() {
        // The wire contract with hippius_s3/writer/landed.py. These three field names are the
        // whole protocol; renaming one on either side silently drops every message, because a
        // parse failure is discarded by design.
        let raw = format!(r#"{{"object_id":"{UUID}","version":7,"part_number":3}}"#);
        let part = serde_json::from_str::<LandedMessage>(&raw).unwrap().into_part().unwrap();

        assert_eq!(part.object().as_str(), UUID);
        assert_eq!(part.version().get(), 7);
        assert_eq!(part.part().get(), 3);
    }

    #[test]
    fn a_non_uuid_object_id_is_dropped_rather_than_recorded() {
        // `object_id` reaches a SQL bind and a filesystem path, so an unvalidated value is both
        // a correctness and a traversal concern. Dropping is safe: the reconciler still finds
        // the part on disk, so a bad message costs latency, never the part.
        let raw = r#"{"object_id":"../../etc/passwd","version":1,"part_number":1}"#;
        assert!(serde_json::from_str::<LandedMessage>(raw).unwrap().into_part().is_none());
    }

    #[test]
    fn a_message_missing_a_field_does_not_parse() {
        let raw = format!(r#"{{"object_id":"{UUID}","version":1}}"#);
        assert!(serde_json::from_str::<LandedMessage>(&raw).is_err());
    }

    #[test]
    fn garbage_does_not_parse() {
        assert!(serde_json::from_str::<LandedMessage>("not json").is_err());
    }
}
